server/hap2/haplib.py

```diff
@@ -304,3 +308,166 @@ def error(self, error_code, response_id):
+        if need_id and pm.message_id is None:
+            pm.error_code = ERR_CODE_INVALID_PARAMS
+            pm.error_message = "Not found: id"
+            return pm
+
+        if should_reply:
+            return pm
+
+        pm.error_code, pm.error_message = \
+          Utils.validate_arguments(pm.message_dict)
+        if pm.error_code is not None:
+            return pm
+
+        return pm
+
+    @staticmethod
+    def _convert_string_to_dict(json_string):
+        try:
+            json_dict = json.loads(json_string)
+        except ValueError:
+            return (ERR_CODE_PARSER_ERROR, None)
+        else:
+            return (None, json_dict)
+
+    @staticmethod
+    def is_allowed_procedure(procedure_name, allowed_procedures):
+        if procedure_name in allowed_procedures:
+            return
+        else:
+            return ERR_CODE_METHOD_NOT_FOUND
+
+    @staticmethod
+    def validate_arguments(json_dict):
+        args_dict = PROCEDURES_DEFS[json_dict["method"]]["args"]
+        for arg_name, arg_value in args_dict.iteritems():
+            try:
+                type_expect = type(json_dict["params"][arg_name])
+                type_actual = type(arg_value["type"])
+                if type_expect != type_actual:
+                    msg = "Argument '%s': unexpected type: exp: %s, act: %s" \
+                          % (arg_name, type_expect, type_actual)
+                    return (ERR_CODE_INVALID_PARAMS, msg)
+            except KeyError:
+                if arg_value["mandatory"]:
+                    msg = "Missing a andatory paramter: %s" % arg_name
+                    return (ERR_CODE_INVALID_PARAMS, msg)
+        return (None, None)
+
+    @staticmethod
+    def generate_request_id(component_code):
+        assert component_code <= 0x7f, \
+               "Invalid component code: " + str(component_code)
+        req_id = random.randint(1, 0xffffff)
+        req_id |= component_code << 24
+        return req_id
+
+    @staticmethod
+    def translate_unix_time_to_hatohol_time(float_unix_time):
+        return datetime.strftime(datetime.fromtimestamp(float_unix_time), "%Y%m%d%H%M%S.%f")
+
+    @staticmethod
+    def translate_hatohol_time_to_unix_time(hatohol_time):
+        date_time = time.strptime(hatohol_time, "%Y%m%d%H%M%S.%f")
+        return int(time.mktime(date_time))
+
+    @staticmethod
+    def optimize_server_procedures(valid_procedures_dict, procedures):
+        for name in valid_procedures_dict:
+            valid_procedures_dict[name] = False
+
+        for procedure in procedures:
+            valid_procedures_dict[procedure] = True
+
+    #This method is created on the basis of getting same number of digits under the decimal.
+    @staticmethod
+    def get_biggest_num_of_dict_array(array, key):
+        last_info = None
+
+        for target_dict in array:
+            if isinstance(target_dict[key], int):
+                break
+            digit = int()
```

Comment:
The "digit" will be initialized each time so that the last value is always used.